docs/rfcs/017-timeline-data-management.md

```diff
@@ -0,0 +1,391 @@
+
+- change the approach to locking synchronization
+
+A number of locks in the timeline seem to be used to coordinate gc, compaction tasks and related processes.
+It should be done in a task manager or other place, external to the timeline.
+
+Timeline contents still needs to be synchronized, considering the task work, so fields like `latest_gc_cutoff_lsn: RwLock<Lsn>` are expected to stay for that purpose, but general amount of locks should be reduced.
+
+### Putting it all together
+
+If the proposal bullets applied to the diagrams above, the state could be represented as:
+
+![timeline timeline tenant state](./images/017-timeline-data-management/proposed_timeline_tenant_state.svg)
+
+The reorders aim to put all tasks into separated modules, with strictly defined interfaces and as less knowledge about other components, as possible.
+This way, all timeline data is now in the `data_storage`, including the GC, walreceiver, `RemoteTimelineIndex`, `LayerMap`, etc. with some API to get the data in the way,
+more conveninent for the data sync system inside.
+So far, it seems that a few maps with `Arc<RwLock<SeparateData>>` with actual data operations added inside each `SeparateData` struct, if needed.
+
+`page_cache` is proposed to placed into the same `data_storage` since it contains tenant timelines' data: this way, all metadata and data is in the same struct, simplifying things with Rust's borrow checker and allowing is to do care about data sync withing a single module.
+
+`task_manager` is related to data storage and manages all tenant and timeline tasks, manages shared resources (runtimes, threadpools, etcd connection, etc.) and synchronizes tasks.
+All locks such as `gc_cs` belong to this module tree, as primitives inherently related to the task synchronization.
+Tasks have to access timelines and their metadata, but should do that through `data_storage` API and similar.
+
+`task_manager` should (re)start, stop and track all tasks that are run in it, selecting an appropriate runtime depending on a task kind (we have async/sync task separation, CPU and IO bound tasks separation, ...)
+Some locks such as `layer_removal_cs` one are not needed, if the only component that starts the tasks ensures they don't run concurrently.
+
+`LayeredTimeline` is still split into two parts, more high-level with whatever promitives needed to sync its state, and the actual state storage with `LayerMap` and other low level entities.
+Only `LayerMap` knows what storage it's layer files are taken from (inmem, local FS, etc.), and it's responsible for synchronizing the layers when needed, as also reacting to sync events, successful or not.
+
+Last but not least, `tenant config file` has to be backed into a remote storage, as tenant-specific information for all timelines.
```

Comment:
I don't think we need that. Tenant config is stored in the console anyway. Should be ok to require config in attach call and then keep it on pageserver disk. Console anyway needs to call into pageserver to apply changed config so I dont have a clear reason to duplicate this in remote storage. WDYT?

Reply:
The whole story with this file seems unfinished and I'm not aware of any plans, alas.
I like the idea in general, since requires less code to handle yet not sure how other components are working with this.

Your proposal conflicts with the `PUT` pageserver HTTP API method for changing the config at least, but might be ok anyway, if the updated config is stored elsewhere.

Reply:
> Tenant config is stored in the console anyway

Small note, it isn't, at least now, and it wasn't planed, AFAIK. The only thing that console (control-plane later) knows about tenant is that it exists and its id.

If there are some knobs that control-plane can adjust and we really want to do that, then we can save tenant config returned after creation call in the console db. Yet, there is at least one problem I see with it right now -- if it's some old tenant, we could add new options on the pageserver side already and they have some defaults, but in the console db it's still the same and old. So we need some duplicative logic in the console that knows about all per-tenant options and their defaults for that case.

Regarding the keeping a per-tenant config in the remote storage, I think it may still have sense. Imagine that pageserver instance was completely lost, so we restore it from the S3. If S3 contains all the required information it can recover on its own. Otherwise, we loose all tenant configs, so console need to explicitly attach all previous tenants back to the freshly created pageserver? IIUC, that's what @LizardWizzard suggests

Sorry, if I completely missed the point, just stumbled upon the discussion and wanted to clarify the current state

Reply:
See https://github.com/neondatabase/neon/pull/1504#issuecomment-1099106458. Tenant config includes PITR settings. Console anyway needs to store that. There is no need to keep default config, console can only pass settings that override the defaults. Whatever is more convenient

In case of pageserver loss restore is supposed to happen via creation of new pageserver and attaching tenants to it. My understanding is that attach call should accept config as an input. 

So I dont see any problems with this schema